snuba/web/split.py

```diff
@@ -128,61 +159,89 @@ def time_split(dataset, request: Request, *args, **kwargs):
-        # The query function may mutate the request body during query
-        # evaluation, so we need to copy the body to ensure that the query has
-        # not been modified by the time we're ready to run the full query.
-        minimal_request = copy.deepcopy(request)
-        minimal_request.query.set_selected_columns(column_split_spec.get_min_columns())
-        result = query_func(dataset, minimal_request, *args, **kwargs)
-        del minimal_request
-
-        if result.result["data"]:
-            request = copy.deepcopy(request)
-
-            event_ids = list(
-                set(
-                    [
-                        event[column_split_spec.id_column]
-                        for event in result.result["data"]
-                    ]
-                )
-            )
-            request.query.add_conditions(
-                [(column_split_spec.id_column, "IN", event_ids)]
-            )
-            request.query.set_offset(0)
-            request.query.set_limit(len(event_ids))
-
-            project_ids = list(
-                set(
-                    [
-                        event[column_split_spec.project_column]
-                        for event in result.result["data"]
-                    ]
-                )
-            )
-            request.extensions["project"]["project"] = project_ids
-
-            timestamp_field = column_split_spec.timestamp_column
-            timestamps = [event[timestamp_field] for event in result.result["data"]]
-            request.extensions["timeseries"]["from_date"] = util.parse_datetime(
-                min(timestamps)
-            ).isoformat()
-            # We add 1 second since this gets translated to ('timestamp', '<', to_date)
-            # and events are stored with a granularity of 1 second.
-            request.extensions["timeseries"]["to_date"] = (
-                util.parse_datetime(max(timestamps)) + timedelta(seconds=1)
-            ).isoformat()
-
-        return query_func(dataset, request, *args, **kwargs)
-
-    return wrapper
+        if (
+            not query.get_limit()
+            or query.get_groupby()
+            or query.get_aggregations()
+            or not query.get_selected_columns()
+        ):
+            return None
+
+        total_col_count = len(query.get_all_referenced_columns())
+        minimal_query = copy.deepcopy(query)
+        minimal_query.set_selected_columns(
+            [self.__id_column, self.__project_column, self.__timestamp_column]
+        )
+
+        if total_col_count <= len(minimal_query.get_all_referenced_columns()):
+            return None
+
+        result = runner(minimal_query, request_settings)
+        del minimal_query
+
+        if not result.result["data"]:
+            return None
+
+        # Making a copy just in case runner returned None (which would drive the execution
+        # strategy to ignore the result of this splitter and try the next one).
+        query = copy.deepcopy(query)
```

Comment:
This should never be needed. I added it defensively just in case we add another condition that returns None and we forget that we mutated the query.